crates/rpc-types-engine/src/envelope.rs

```diff
@@ -42,10 +42,13 @@ impl ssz::Encode for OpExecutionPayloadEnvelope {
     }
 
     fn ssz_append(&self, buf: &mut Vec<u8>) {
-        // Write parent beacon block root
-        match &self.parent_beacon_block_root {
-            Some(root) => buf.extend_from_slice(root.as_slice()),
-            None => buf.extend_from_slice(&[0u8; 32]),
+        // Write parent beacon block root only if the payload is not a v1 or v2 payload.
+        // <https://specs.optimism.io/protocol/rollup-node-p2p.html#block-encoding>
+        if !matches!(self.payload, OpExecutionPayload::V1(_) | OpExecutionPayload::V2(_)) {
+            match &self.parent_beacon_block_root {
+                Some(root) => buf.extend_from_slice(root.as_slice()),
+                None => buf.extend_from_slice(&[0u8; 32]),
+            }
         }
```

Comment:
This works, but let's mandate that the `parent_beacon_block_root` is present for V3/V4 and panic if it's present on V1/V2 to prevent implicit footguns.

```suggestion
        match self.payload {
            OpExecutionPayload::V1(_) | OpExecutionPayload::V2(_) => {
                if self.parent_beacon_block_root.is_some() {
                    panic!("Parent beacon block root present on pre-V3 execution payload envelope");
                }
            }
            OpExecutionPayload::V3(_) | OpExecutionPayload::V4(_) => {
                match &self.parent_beacon_block_root {
                    Some(root) => buf.extend_from_slice(root.as_slice()),
                    None => panic!("Parent beacon block root not present on V3 or greater execution payload envelope"),
                }
            }
        }
```